astropy/time/core.py

```diff
@@ -1476,15 +1479,26 @@ def __new__(cls, val, val2=None, format=None, scale=None,
                 location=None, copy=False):
 
         # Because of import problems, this can only be done on
-        # first call of Time.
-        global _LEAP_SECONDS_CHECKED
+        # first call of Time. The initialization is complicated because
+        # update_leap_seconds uses Time.
+        # In principle, this may cause wrong leap seconds in
+        # update_leap_seconds itself, but since expiration is in
+        # units of days, that is fine.
+        global _LEAP_SECONDS_CHECKED, _LEAP_SECONDS_CHECK_STARTED
         if not _LEAP_SECONDS_CHECKED:
-            # *Must* set to True first as update_leap_seconds uses Time.
-            # In principle, this may cause wrong leap seconds in
-            # update_leap_seconds itself, but since expiration is in
-            # units of days, that is fine.
-            _LEAP_SECONDS_CHECKED = True
-            update_leap_seconds()
+            with _LEAP_SECONDS_LOCK:
+                # There are three ways we can get here:
+                # 1. First call.
+                # 2. Re-entrant call. Then _LEAP_SECONDS_CHECK_STARTED will
+                #    already be true, and we skip the initialisation and
+                #    don't worry about leap second errors.
+                # 3. Another thread which raced with the first call. The
+                #    first thread has relinquished the lock to us, so
+                #    initialization is complete.
+                if not _LEAP_SECONDS_CHECK_STARTED:
```

Comment:
I like the clear description! A possibility perhaps: instead of adding a new variable, `_LEAP_SECONDS_CHECKED` could go from `False` to `None` to `True`. Though perhaps it is just less clear for no benefit.

Reply:
In a similar vein, I find the naming of `_LEAP_SECONDS_CHECK_STARTED` a little misleading, as once it is set to `True`, it never changes back to `False` when it has ended (which to me is the opposite of started). So either re-use `_LEAP_SECONDS_CHECKED` as @mhvk said above, or consider renaming it.

Reply:
False/None/True smells a bit too much like [this](http://thedailywtf.com/articles/What_Is_Truth_0x3f_), but I had considered using an enum for the three states. That will simplify the test code a bit since only one variable needs to be poked.

Reply:
Done (with an enum).